src/goto-instrument/contracts/dynamic-frames/dfcc_loop_utils.cpp

```diff
@@ -0,0 +1,514 @@
+/*******************************************************************\
+
+Module: Dynamic frame condition checking for loop contracts
+
+Author: Qinheping Hu, qinhh@amazon.com
+
+\*******************************************************************/
+
+#include "dfcc_loop_utils.h"
+
+#include <util/c_types.h>
+#include <util/fresh_symbol.h>
+#include <util/pointer_expr.h>
+
+#include <goto-instrument/contracts/utils.h>
+
+const irep_idt ID_loop_nesting_id = "loop-id";
+const irep_idt ID_loop_instr_type = "loop-instr-type";
+const irep_idt ID_loop_instr_body = "body";
+const irep_idt ID_loop_instr_head = "head";
+const irep_idt ID_loop_instr_latch = "latch";
+const irep_idt ID_loop_instr_exiting = "exiting";
+
+void tag_loop_instr(
+  goto_programt::instructiont::targett &target,
+  const irep_idt &instr_type,
+  const std::size_t loop_nesting_id)
+{
+  irept &irep = target->source_location_nonconst();
+  irep.set(ID_loop_instr_type, instr_type);
+  irep.set(ID_loop_nesting_id, loop_nesting_id);
+}
+
+irep_idt get_loop_instr_type(const goto_programt::instructiont::targett &target)
+{
+  irept &irep = target->source_location_nonconst();
+  return irep.get(ID_loop_instr_type);
+}
+
+std::size_t
+get_loop_nesting_id(const goto_programt::instructiont::targett &target)
+{
+  PRECONDITION(!get_loop_instr_type(target).empty());
+  irept &irep = target->source_location_nonconst();
```

Comment:
Use `source_locationt` or `auto` instead of `irept`. (Yes, we inherit from that, but it's an implementation detail.)